src/input/composite_device/mod.rs

```diff
@@ -1479,6 +1479,23 @@ impl CompositeDevice {
         let path = device.devnode();
         let id = device.get_id();
 
+        let mut freed_effect_ids = Vec::new();
+
+        for (effect_id, source_map) in self.ff_effect_id_source_map.iter_mut() {
+            if source_map.remove(&id).is_some() {
+                log::debug!("Removed source {id} from effect {effect_id}");
+            }
+            if source_map.is_empty() {
```

Comment:
Add Comment
`// If this was the last user of the effect, stage the effect for removal`